### DISCUSSION TITLE
VA SIGMET with no volcano coordinates

### DISCUSSION THREAD

**Discussion by ernstdevreede** [wmo-im/iwxxm#275] [Q&A] [answered]
From Annex 3 it appears that in VA SIGMET's the volcano's coordinates are optional for TAC.

```
[VA ERUPTION]
[MT nnnnnnnnnn]
[PSN Nnn[nn] or Snn[nn]
Ennn[nn]
VA CLD

```
In IWXXM (>=3.0) however these coordinates are mandatory in the EruptingVolcano type, so it's not possible to generate a legal IWXXM without coordinates.

Does this mean that the volcano coordinates are mandatory in both TAC and IWXXM?

**Comment by mgoberfield** [comment]
Hello,

In cases when the source volcano coordinates are not known, you can set element <metce:position> attribute 'nilReason' to the [appropriate value](http://codes.wmo.int/common/nil).

For example:
```
<iwxxm:eruptingVolcano>
    <metce:EruptingVolcano gml:id="uuid.f59689d4-2376-475c-921f-91bea3f63cfb">
        <metce:name>UNKNOWN</metce:name>
        <metce:position nilReason="http://codes.wmo.int/common/nil/unknown"/>
        <metce:eruptionDate>2021-11-01T12:00:00Z</metce:eruptionDate>
    </metce:EruptingVolcano>
</iwxxm:eruptingVolcano>
```

Respectfully,

mark

**Comment by ernstdevreede** [comment]
Heelo Mark,

Thanks for your response.
I tried this earlier, but ran into a validation error.
```
My SIGMET (IWXXM3.0) had:
   <iwxxm:eruptingVolcano>
      <metce:Volcano gml:id="uuid.90e40563-23a0-49f3-a2c2-c115e2fed05d">
         <metce:name>GRIMSVOTN</metce:name>
         <metce:position nilReason="http://codes.wmo.int/common/nil/unknown" xsi:nil="true"/>
      </metce:Volcano>
   </iwxxm:eruptingVolcano>

```
This results in 2 validation errors:
```

iss:XML Schema validation error at line 141, column 94:cvc-elt.3.1: Attribute 'http://www.w3.org/2001/XMLSchema-instance,nil' must not appear on element 'metce:position', because the {nillable} property of 'metce:position' is false.===org.xml.sax.SAXParseException; lineNumber: 141; columnNumber: 94; cvc-elt.3.1: Attribute 'http://www.w3.org/2001/XMLSchema-instance,nil' must not appear on element 'metce:position', because the {nillable} property of 'metce:position' is false.
iss:Failed Schematron assertation: VolcanicAshSIGMET.VolcanicAshSIGMET-6: If a geometry is defined for iwxxm:eruptingVolcano//metce:position by providing @srsName, @srsDimension must equal to 2 and @aixsLabels must be non-empty===null

```

I beieve that if there is a nilReason, there also has to be the xsi:nil="true" ?
Is that not correct?

**Reply by mgoberfield** [reply]
Hello,

You are correct: 'xsi:nil=true' attribute **should** be there, but the METCE schema for EruptingVolcano/position element does not include the nillable='true' attribute, so the use of xsi:nil is flagged as you've pointed out.

For now, remove the xsi:nil attribute, then it should validate (it does for me in OxygenXML).

I've been working this [issue](https://github.com/wmo-im/iwxxm-modelling/issues/14) in the IWXXM schemas. The task team will look into the possibility of fixing the METCE schema.

Respectfully,

mark

**Reply by ernstdevreede** [reply]
I tried this, but still run into the same
VolcanicAshsigmet:VolcanicAshSigmet-6 validation problem with

```
   <iwxxm:eruptingVolcano>
      <metce:Volcano gml:id="uuid.90e40563-23a0-49f3-a2c2-c115e2fed05d">
         <metce:name>GRIMSVOTN</metce:name>
         <metce:position nilReason="http://codes.wmo.int/common/nil/unknown"/>
      </metce:Volcano>
   </iwxxm:eruptingVolcano>

```

**Reply by ernstdevreede** [reply]
To me it looks like the validation rule VolcanicAshsigmet:VolcanicAshSigmet-6  makes it impossible to generate a Volcanic Ash SIGMET without a metce:position having all the required elements.
Nilreason with xsi:nil adheres to the xsd schema for metce, but does not validate with the Schematron rules.

**Reply by mgoberfield** [reply]
Hello,

You are missing the ```<eruptionDate>``` element, but I see your point w.r.t. the schematron rule. 

Thank you for the bug report. We will try to get the METCE schema fixed.

I am a bit puzzled because I was able to validate this nilReason case against versions 3.0 and 2021-2 of the schematron rules without errors using both CRUX and oxygenXML applications. I am not a schematron expert so I will bring this case up with him at our next meeting to discuss these results.

Respectfully,

mark

**Reply by ernstdevreede** [reply]
Hi Mark,

My question might not have been clear enough. I'm discussing the structure of iwxxm:eruptingVolcano in VA SIGMET's, not the volcano description in VA advisories.

I don't see the eruptionDate at all in the eruptingVolcano element of the SIGMET xsd.

I must say the VolcanicAshSIGMET and the VolcanicAsh Advisory are a bit confusing in specifying the volcano information:
VolcanicAshSigmet:
```
<iwxxm:eruptingVolcano>
  <metce:Volcano>
    <metce:name>...</metce:name>
    <metce:position>...</metce:position>
  </metce:Volcano>
</iwxxm:eruptingVolcano>
```
whereas Volcanic Ash Advisory has:
```
<iwxxm:volcano>
  <metce:EruptingVolcano>
    <metce:name>...
    <metce:position>....
    <metce:eruptionDate>...
  </metce:EruptingVolcano>
</iwxxm:volcano>
```

So the SIGMET has no eruptionDate, but can't have an empty position element (with a nilReason)

Perhaps you didn't get any validation errors, because you were testing Volcanic Ash advisories?

**Comment by mgoberfield** [comment]
Hello,

Thank you for the note.  For VolcanicAshSIGMET the ```<eruptingVolcano>``` element can be either of two types:

![image](https://user-images.githubusercontent.com/7852621/139718278-ad2cd79e-99fa-496e-86db-90f4474f358a.png)

The difference between the two types is the addition of the eruption date in ```<metce:EruptingVolcano>```, and I failed to catch your use of ```<metce:Volcano>```, so thank you for the correction.

However, in either case, I still don't get the schematron error if nilReason is used in the ```<metce:position>``` element.

Respectfully,

mark

**Reply by ernstdevreede** [reply]
I definitely do get these Schematron errors (also in crux based Schematron validators), so not only is it impossible to generate these SIGMET IWXXM's but if I could it would probably never be decoded correctly because it will fail validation on input too.

For now I think the only solution to go on is that we'll have to assume that all VA SIGMETs carry the volcano coordinates; if not we don't generate IWXXM. Translation from TAC would also fail if the coordinates are missing.

I'm wondering if others may have thought up a better work-around.

@mgoberfield Thanks a lot for your answers

Regards,

Ernst de Vreede

**Comment by blchoy** [comment]
Thanks @mgoberfield for following up on this matter.  I have no problem validating with both schemas and schematron for the sample VA SIGMET modified to have no erupting volcano location (see https://gist.github.com/blchoy/cfa2d8bb95e53c14e55a9e0b971ac7db).

Checking the schematron rule VolcanicAshSIGMET.VolcanicAshSIGMET-6 it should not fire when the XML instance has no including elements under `<metce:position>`.  Could you confirm this rule was triggered by the XML instance fragment you mentioned?

**Comment by ernstdevreede** [comment]
I'm pretty sure the same construct doesn't validate in an IWXXM file with IWXXM3.0 rules.
It doesn't validate in my Java code with the VolcanicAshSIGMET.VolcanicAshSIGMET-6 error.
An example of a not validating IWXXM: https://gist.github.com/ernstdevreede/dd5c3afd7c140f4286b14430671e8c6d

Does this SIGMET validate on your side?

**Comment by blchoy** [comment, marked as answer]
Now I know what happened.  Mark and I may have overlooked that you are making reference to IWXXM 3.0 (the next version IWXXM 2021-2 will be out within 2 weeks and my example is on this version).  For VolcanicAshSIGMET.VolcanicAshSIGMET-6 in IWXXM 3.0 there was a known bug (see https://github.com/wmo-im/iwxxm/wiki/Confirmed-Issues-That-May-Affect-Operational-Versions) which had been fixed in IWXXM 3.0-dev.  Unfortunately this version has never been made available officially on schemas.wmo.int.  To get a copy of the fixed iwxxm.sch you can visit https://github.com/wmo-im/iwxxm/releases/tag/v3.0-dev.
